### ISSUE TITLE
proposal: go/types: add Import, Export functions

### ISSUE THREAD

**Issue by adonovan** [golang/go#69491]
[Status: WORK IN PROGRESS. Do not review yet.]

**Background:** The package [golang.org/x/tools/go/gcexportdata](https://pkg.go.dev/golang.org/x/tools/go/gcexportdata) provides two functions, Write and Read. As described in https://github.com/golang/go/issues/68898, they serve two purposes:

1. a serialization mechanism for types.Package values. They assume that writer and reader are compiled from the same source code; there is no protocol stability.
2. a mechanism to read export data written by last "tip + 2" releases of the Go compiler.

These two purposes are quite independent, and the actual protocols they use are unrelated. (The first uses "indexed" form, the second "unified".) Much of the challenge of maintaining go/types and x/tools/go/gcexportdata comes from version skew: witness the number of delicate changes to both made recently to support materialized Alias types, and then the addition of type parameters to aliases.

**Proposal:** We propose to add the read and write functions to go/types itself:

```go
package types // "go/types"

// Export returns a serialized form of pkg that can be read by the same version of [Import].
// The encoding is not specified and may change at any time.
func Export(fset *token.FileSet, pkg *Package, opts ExportOptions) ([]byte, error)

type ExportOptions struct {
    Deep bool // encode information about dependencies too
}

// Import returns the package described by data,
// which must have been produced by the same version of [Export].
// It calls importDeps exactly once to obtain all needed direct dependencies (perhaps in parallel).
func Import(fset *token.FileSet, data []byte, importDeps importDepsFunc, path string, opts ImportOptions) (*Package, error)

type ImportOptions struct { /* for future use */ }
```

This API is a synthesis of gcexportdata and internal/gcimporter.I{Im,Ex}portShallow, which are used by gopls for "shallow" encodings.

Open questions:
- the implementation of Export and Import depend on objectpath. How would this work in go/types?

@griesemer @findleyr @timothy-king

**Comment by findleyr** [comment]
See also #52163.

The export data format has a versioning header precisely for this reason. Presumably if the version has not changed, go/types.Import should be able to read the export data produced by an older version of go/types. On the other hand, if the version _changes_, we'll need to snapshot the logic of the older version in x/tools. Therefore, x/tools will have already interpretted the version header before it calls into go/types. Is you expectation that the bytes passed to go/types will be stripped of this header?

**Comment by adonovan** [comment]
> Is your expectation that the bytes passed to go/types will be stripped of this header?

No, for better or worse the existing gcexportdata.Read function must continue to support files produced both by the compiler (2 releases + tip) and by gcexportdata.Write (which will become a wrapper around types.Export), so we would still need some way to detect the format. I propose that we pick a new version byte (in addition to `[vcdiu]`) that means "the private version namespace of types.{Export.Import}", and use this to dispatch to Import. Import and Export can then do whatever they want with the payload, including having a second version field that we bump as often as we like (e.g. on every change), so that we can report the error "Import called on a file produced by a different source version of Export".

**Comment by adonovan** [comment]
@cherrymui: I think the problems of this proposal might equally apply to a forked export data format that is written by `go list -export` and read and written by functions in x/tools: the read function would have to support older versions of go list. It may still be worth doing though.

**Comment by adonovan** [comment]
Was discussing https://github.com/golang/go/issues/79427 today with @mrkfrmn.)

Two non-objections to this proposal at a high level:

 1. One of the reasons claimed in the past for x/tools having its own export data reader was so that it could be upgraded to read new formats in advance of a toolchain release, allowing applications to be redeployed also in advance of the toolchain release and thus smoothly continue to operate with older and newer toolchains across the transition. But I don't think this argument is valid in general, because a new toolchain release often introduces whole new nouns into the vocabulary (e.g. generics, aliases), and there is simply no way that an application built with an older type checker can express these concepts. So I don't see that as an objection to this proposal.

2. Another reason for x/tools to have its own reader/writer is so that it can add bells and whistles not needed by the standard implementation, for example the ability to export "shallow" export data (which gopls uses), or "[bundles](https://github.com/golang/go/issues/69573)" for export data, used AFAICT only by x/exp/cmd/apidiff. But if an application needs a nonstandard reader/writer, then it can fork the standard one. That's exactly what x/tools has been doing all this time, after all. If we could abolish bundle and upstream the "shallow" feature then the need for the fork would be temporary.

This is analogous to, say, an application forking go/parser to do a better job of error recovery. It helps that go/parser and go/format are separate packages from go/ast as they "keep the ast package honest" about exposing a complete set of constructors (for parser) and accessors (for format). Perhaps that's an argument for putting the Export and Import functions in a separate package such as go/types/export.

But it is clear that the reader/writer functionality of go/types would need to follow a protocol versioning scheme in order that applications build with one toolchain can read files produced by `go list -export` using a (somewhat) older toolchain. So the version text should read something like this:
```go
// Export returns a serialized form of pkg that can be read by [Import].
// The encoding is not specified and may change at any time.
// However, we maintain "2 releases + tip" compatibility per issue #68898.
func Export(fset *token.FileSet, pkg *Package, opts ExportOptions) ([]byte, error)

// Import returns the package described by data,
// which must have been produced by a recent implementation of [Export].
// It calls importDeps at most once to obtain all needed direct dependencies (perhaps in parallel).
func Import(fset *token.FileSet, data []byte, importDeps importDepsFunc, path string, opts ImportOptions) (*Package, error)
```
